tests/Query/Filter/DueDateField.test.ts

```diff
@@ -59,6 +75,13 @@ describe('explain due date queries', () => {
         const filterOrMessage = new DueDateField().createFilterOrErrorMessage('due 2023-01-02');
         expect(filterOrMessage).toHaveExplanation('due date is on 2023-01-02 (Monday 2nd January 2023)');
     });
+
+    it('in current week/month/year', () => {
```

Comment:
> Nicer still would be separate the data from the tests:
> 
> ```ts
>     describe('should explain date ranges', () => {
>         it.each([
>             [
>                 'done in current week',
>                 'done date is between 2022-01-09 (Sunday 9th January 2022) and 2022-01-15 (Saturday 15th January 2022)',
>             ],
>             [
>                 'done in current month',
>                 'done date is between 2022-01-01 (Saturday 1st January 2022) and 2022-01-31 (Monday 31st January 2022)',
>             ],
>             [
>                 'done in current year',
>                 'done date is between 2022-01-01 (Saturday 1st January 2022) and 2022-12-31 (Saturday 31st December 2022)',
>             ],
>         ])('explains "%s" as "%s"', (filter: string, expectedExpanation: string) => {
>             const filterOrMessage = new DoneDateField().createFilterOrErrorMessage(filter);
>             expect(filterOrMessage).toHaveExplanation(expectedExpanation);
>         });
>     });
> ```
> 
> This would be easy to extend as more range-based capabilities are added.

I tried following this path, but the amount of code generated is pretty big, since the test input date is repeated, but simple (Possibility for refactoring, but I don't see the good one).

For instance: https://github.com/obsidian-tasks-group/obsidian-tasks/pull/1638/commits/b2a00a6af259a93a283da75d6e92f139e9dc1c93

The other path I tried is creating a something like a "tester". The code is much more compact, but the testers become more complicated once they expand to other filters like `happens` and `starts`. Also calling `expect()` outside of the test itself eg in a helper lib might be a bad idea.

Here is an example: https://github.com/obsidian-tasks-group/obsidian-tasks/pull/1638/commits/8cfcb2f7e546d2846da6410887d4f078bff4e8b9

I have a slight preference for the second way but I'd like your opinion first on the best path you see fit.

Reply:
Hi, I've only had limited time to look at this, but I think I understand what's going on. It's great progress.

In https://github.com/obsidian-tasks-group/obsidian-tasks/commit/b2a00a6af259a93a283da75d6e92f139e9dc1c93, the tests are very clear: the inputs and expected results are stated directly in the code.

In https://github.com/obsidian-tasks-group/obsidian-tasks/commit/8cfcb2f7e546d2846da6410887d4f078bff4e8b9, there is no information in the test at all (`tests/Query/Filter/DoneDateField.test.ts`) about what is going on.

I think I mentioned in the code review that being able to reason about the behaviour of the tests was essential.

For this reason, https://github.com/obsidian-tasks-group/obsidian-tasks/commit/b2a00a6af259a93a283da75d6e92f139e9dc1c93 is a clear winner here.

---

But I think there's a bit more that is worth saying...

I think you are probably worried about the same test values being duplicated across due, scheduled, starts etc...

But there is no need to repeat all the detail tests in all of those, because we know that the implementation is shared between them.

For example, if you have a look at this 2 comments in this existing test section, it explains why the tests are minimal, and says where to look for the more detailed ones...

https://github.com/obsidian-tasks-group/obsidian-tasks/blob/8f3612552c5f8605e36fc0e6971e9b54b430c44b/tests/Query/Filter/StartDateField.test.ts#L32-L51

So if I were writing tests for this PR, I would put the detailed tests (as in https://github.com/obsidian-tasks-group/obsidian-tasks/commit/b2a00a6af259a93a283da75d6e92f139e9dc1c93) in just `DueDateField.test.ts` and `HappensDateField.test.ts` - because they use different implementations for now...

And then in the other ones that inherit `DateField`, I would put one or two tests only, just to prove that the respective classes do use the `DateField` logic.

Reply:
> I tried following this path, but the amount of code generated is pretty big, since the test input date is repeated,

Ooooohh - I think I just understood what you mean about the repeated date, @ilandikov. I had presumed that you had added the text version of the dates to make the tests self-documenting, but it is not necessary, and opens up the possibility of inconsistent values being given.

In case you didn't know, you can do this - which I feel is still just as clear, but easier to add new tests:

```ts
    it.each([
        // Week
        ['happens in current week', '2022-01-09', false],
        ['happens in current week', '2022-01-10', true],
        ['happens in current week', '2022-01-16', true],
        ['happens in current week', '2022-01-17', false],

        // Month
        ['happens in current month', '2021-12-31', false],
        ['happens in current month', '2022-01-01', true],
        ['happens in current month', '2022-01-31', true],
        ['happens in current month', '2022-02-01', false],

        // Year
        ['happens in current year', '2021-12-31', false],
        ['happens in current year', '2022-01-01', true],
        ['happens in current year', '2022-12-31', true],
        ['happens in current year', '2023-01-01', false],
    ])(
```

Reply:
> Yes, I prefer more explicit dates for tests, this makes it more self-evident what's going on (Especially for the week tests). I'm just copy-pasting those, so I'm not bothered by the lengths of the dates. I'd keep them as they are (longer version).

The longer version adds an overhead for anyone adding new dates in future, though. 

Every reviewer has to check if the dates are consistent. And do you you know what the behaviour would be in the date parser if the two halves of the date differed. 

And maintainable tests matter a lot too!

The ones where the number-only date is unclear is the week ones, so I would just add a comment after dates for those, like 

2933-11-22, // Wednesday

Reply:
> The ones where the number-only date is unclear is the week ones, so I would just add a comment after dates for those, like

Brilliant! Will do =)

Reply:
> I'm just copy-pasting those

yeah, so for future reference for test writing (no need to change it here) having a variety of test data it super valuable, for catching corner case problems. 

For example, a test of ‘this month’ in a February in a leap year would be useful, but the fact that you have to copy tests makes that harder.